File: frontend/src/types/sitrep.ts
/**
 * SITREP-related type definitions
 */

export enum SitrepPriority {
  ROUTINE = 'routine',
  PRIORITY = 'priority',
  IMMEDIATE = 'immediate',
  FLASH = 'flash',
}

export enum SitrepClassification {
  UNCLASSIFIED = 'unclassified',
  CONFIDENTIAL = 'confidential',
  SECRET = 'secret',
  TOP_SECRET = 'top_secret',
}

export enum SitrepSource {
  VOICE = 'voice',
  MANUAL = 'manual',
  AUTO = 'auto',
}

export interface Entity {
  type: string
  value: string
  confidence: number
}

export interface Sitrep {
  id: number
  title: string
  report_time: string
  location?: string
  latitude?: number
  longitude?: number
  unit?: string
  reporter?: string

  // SITREP sections
  situation?: string
  mission?: string
  execution?: string
  admin_logistics?: string
  command_signal?: string

  // Extracted data
  entities?: Entity[]
  source: string
  audio_transcript?: string
  metadata?: Record<string, any>

  priority: SitrepPriority
  classification: SitrepClassification

  created_at: string
  updated_at: string
}

export interface SitrepCreate {
  title: string
  location?: string
  latitude?: number
  longitude?: number
  unit?: string
  reporter?: string
  situation?: string
  mission?: string
  execution?: string
  admin_logistics?: string
  command_signal?: string
  priority?: SitrepPriority
  classification?: SitrepClassification
  source?: SitrepSource
  audio_transcript?: string
}

export interface SitrepUpdate {
  title?: string
  situation?: string
  mission?: string
  execution?: string
  admin_logistics?: string
  command_signal?: string
  priority?: SitrepPriority
}

export interface VoiceDebriefingRequest {
  audio_data: string
  audio_format?: string
  reporter?: string
  location?: string
}

export interface VoiceDebriefingResponse {
  sitrep: Sitrep
  transcript: string
  entities: Entity[]
  processing_time: number
  confidence: number
}

export interface SitrepGenerationRequest {
  text_input: string
  include_entities?: boolean
  auto_classify?: boolean
}

export interface SitrepGenerationResponse {
  sitrep: Sitrep
  entities: Entity[]
  processing_time: number
}

File: backend/app/models/sitrep.py
"""
SITREP data models for API requests and responses
"""

from pydantic import BaseModel, Field
from typing import Optional, Dict, Any, List
from datetime import datetime
from enum import Enum


class SitrepPriority(str, Enum):
    """SITREP priority levels"""
    ROUTINE = "routine"
    PRIORITY = "priority"
    IMMEDIATE = "immediate"
    FLASH = "flash"


class SitrepClassification(str, Enum):
    """SITREP classification levels"""
    UNCLASSIFIED = "unclassified"
    CONFIDENTIAL = "confidential"
    SECRET = "secret"
    TOP_SECRET = "top_secret"


class SitrepSource(str, Enum):
    """SITREP source types"""
    VOICE = "voice"
    MANUAL = "manual"
    AUTO = "auto"


class Entity(BaseModel):
    """Extracted entity from SITREP"""
    type: str  # location, unit, equipment, personnel, time, etc.
    value: str
    confidence: float = Field(ge=0.0, le=1.0)


class SitrepBase(BaseModel):
    """Base SITREP model"""
    title: str
    location: Optional[str] = None
    latitude: Optional[float] = Field(None, ge=-90, le=90)
    longitude: Optional[float] = Field(None, ge=-180, le=180)
    unit: Optional[str] = None
    reporter: Optional[str] = None

    # SITREP sections
    situation: Optional[str] = None
    mission: Optional[str] = None
    execution: Optional[str] = None
    admin_logistics: Optional[str] = None
    command_signal: Optional[str] = None

    priority: SitrepPriority = SitrepPriority.ROUTINE
    classification: SitrepClassification = SitrepClassification.UNCLASSIFIED


class SitrepCreate(SitrepBase):
    """Model for creating a new SITREP"""
    source: SitrepSource = SitrepSource.MANUAL
    audio_transcript: Optional[str] = None


class SitrepUpdate(BaseModel):
    """Model for updating a SITREP"""
    title: Optional[str] = None
    situation: Optional[str] = None
    mission: Optional[str] = None
    execution: Optional[str] = None
    admin_logistics: Optional[str] = None
    command_signal: Optional[str] = None
    priority: Optional[SitrepPriority] = None


class Sitrep(SitrepBase):
    """Complete SITREP model with database fields"""
    id: int
    report_time: datetime
    source: str
    entities: Optional[List[Dict[str, Any]]] = None
    audio_transcript: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class VoiceDebriefingRequest(BaseModel):
    """Request model for voice debriefing"""
    audio_data: str  # Base64 encoded audio
    audio_format: str = "wav"  # wav, mp3, ogg
    reporter: Optional[str] = None
    location: Optional[str] = None


class VoiceDebriefingResponse(BaseModel):
    """Response model for voice debriefing"""
    sitrep: Sitrep
    transcript: str
    entities: List[Entity]
    processing_time: float
    confidence: float


class SitrepGenerationRequest(BaseModel):
    """Request model for automated SITREP generation"""
    text_input: str
    include_entities: bool = True
    auto_classify: bool = True


class SitrepGenerationResponse(BaseModel):
    """Response model for automated SITREP generation"""
    sitrep: Sitrep
    entities: List[Entity]
    processing_time: float
